flink/v1.12/flink/src/test/java/com/netease/arctic/flink/read/hybrid/enumerator/TestArcticSourceEnumerator.java

```diff
@@ -0,0 +1,253 @@
+import org.apache.flink.api.connector.source.SourceEvent;
+import org.apache.flink.api.connector.source.SplitEnumeratorContext;
+import org.apache.flink.api.connector.source.SplitsAssignment;
+import org.apache.flink.metrics.MetricGroup;
+import org.apache.flink.table.data.GenericRowData;
+import org.apache.flink.table.data.RowData;
+import org.apache.flink.table.data.StringData;
+import org.apache.flink.table.data.TimestampData;
+import org.apache.flink.types.RowKind;
+import org.apache.iceberg.flink.FlinkSchemaUtil;
+import org.apache.iceberg.io.TaskWriter;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.io.IOException;
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import java.time.LocalTime;
+import java.time.ZoneOffset;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Callable;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.BiConsumer;
+
+import static org.apache.flink.util.Preconditions.checkState;
+
+public class TestArcticSourceEnumerator extends FlinkTestBase {
+
+  protected KeyedTable testKeyedTable;
+  private final int splitCount = 4;
+  public static final String SCAN_STARTUP_MODE_EARLIEST = "earliest";
+
+  protected static final LocalDateTime ldt =
+    LocalDateTime.of(
+      LocalDate.of(2022, 1, 1),
+      LocalTime.of(0, 0, 0, 0));
+
+  public TestArcticSourceEnumerator() {
+    super(new BasicCatalogTestHelper(TableFormat.MIXED_ICEBERG),
+      new BasicTableTestHelper(true, true));
+  }
+
+  @Before
+  public void init() throws IOException {
+    testKeyedTable = getArcticTable().asKeyedTable();
+
+    //write change insert
+    {
+      TaskWriter<RowData> taskWriter = createKeyedTaskWriter(testKeyedTable, FlinkSchemaUtil.convert(BasicTableTestHelper.TABLE_SCHEMA), false);
+      List<RowData> insert = new ArrayList<RowData>() {{
+        add(GenericRowData.ofKind(
+          RowKind.INSERT, 1, StringData.fromString("john"), ldt.toEpochSecond(ZoneOffset.UTC), TimestampData.fromLocalDateTime(ldt)));
+        add(GenericRowData.ofKind(
+          RowKind.INSERT, 2, StringData.fromString("lily"), ldt.toEpochSecond(ZoneOffset.UTC), TimestampData.fromLocalDateTime(ldt.plusDays(1))));
+        add(GenericRowData.ofKind(
+          RowKind.INSERT, 3, StringData.fromString("jake"), ldt.toEpochSecond(ZoneOffset.UTC), TimestampData.fromLocalDateTime(ldt.plusDays(2))));
+        add(GenericRowData.ofKind(
+          RowKind.INSERT, 4, StringData.fromString("sam"), ldt.toEpochSecond(ZoneOffset.UTC), TimestampData.fromLocalDateTime(ldt.plusDays(3))));
+      }};
+      for (RowData record : insert) {
+        taskWriter.write(record);
+      }
+      commit(testKeyedTable, taskWriter.complete(), false);
+    }
+  }
+
+  @Test
+  public void testMultipleReaderWhenSplitDiscovered() throws Exception {
```

Comment:
How about rename to this:
```suggestion
  public void testReadersNumGreaterThanSplits() throws Exception {
```